okapi-core/src/test/java/org/folio/okapi/util/DepResolutionTest.java

```diff
@@ -190,443 +174,715 @@ public void testLatest(TestContext context) {
-          assertEnable(tml.get(0), mdA100);
-        }));
+    DepResolution.install(map(mdA100, mdB, mdC, mdA110, mdE110), map(mdA100), tml, true);
+    Assert.assertEquals(Json.encodePrettily(tml), 1, tml.size());
+    assertAction(tml, 0, Action.enable, mdA100, mdA100);
   }
 
   @Test
-  public void testUpgradeDifferentProduct(TestContext context) {
+  public void testUpgradeDifferentProduct() {
     List<TenantModuleDescriptor> tml = enableList(mdB);
-    DepResolution.installSimulate(map(mdA100, mdB, mdC, mdA110, mdE100), map(mdA100), tml, false)
-        .onComplete(context.asyncAssertSuccess(res -> {
-          logger.debug("tml result = " + Json.encodePrettily(tml));
-          context.assertEquals(2, tml.size());
-          assertDisable(tml.get(0), mdA100);
-          assertEnable(tml.get(1), mdB);
-        }));
+    DepResolution.install(map(mdA100, mdB, mdC, mdA110, mdE100), map(mdA100), tml, false);
+    Assert.assertEquals(Json.encodePrettily(tml), 2, tml.size());
+    assertDisable(tml, 0, mdA100);
+    assertEnable(tml, 1, mdB);
   }
 
   @Test
-  public void testUpgradeSameProduct(TestContext context) {
+  public void testUpgradeDifferentProductNoFixup() {
+    List<TenantModuleDescriptor> tml = enableList(mdB);
+    OkapiError error = Assert.assertThrows(OkapiError.class,
+        () -> DepResolution.installMaxIterations(map(mdA100, mdB, mdC, mdA110, mdE100), map(mdA100), tml, false, 0));
+    Assert.assertEquals("Multiple modules moduleB-1.0.0, moduleA-1.0.0 provide interface int", error.getMessage());
+  }
+
+  @Test
+  public void testMultipleInterfaces() {
+    List<TenantModuleDescriptor> tml = enableList(mdB,mdA100);
+
+    OkapiError error = Assert.assertThrows(OkapiError.class,
+        () -> DepResolution.install(map(mdA100, mdB, mdC, mdA110, mdE100), map(mdA100), tml, false));
+    Assert.assertEquals("Multiple modules moduleB-1.0.0, moduleA-1.0.0 provide interface int", error.getMessage());
+  }
+
+  @Test
+  public void testUpgradeSameProduct() {
     List<TenantModuleDescriptor> tml = enableList(mdA110);
-    DepResolution.installSimulate(map(mdA100, mdB, mdC, mdA110, mdD100, mdE100), map(mdA100), tml, false)
-        .onComplete(context.asyncAssertSuccess(res -> {
-      logger.debug("tml result = " + Json.encodePrettily(tml));
-      context.assertEquals(1, tml.size());
-      assertUpgrade(tml.get(0), mdA110, mdA100);
-    }));
+    DepResolution.install(map(mdA100, mdB, mdC, mdA110, mdD100), map(mdA100), tml, false);
+    Assert.assertEquals(Json.encodePrettily(tml), 1, tml.size());
+    assertUpgrade(tml, 0, mdA110, mdA100);
+  }
+
+  @Test
+  public void testUpgrade1() {
+    List<TenantModuleDescriptor> tml = enableList(mdD200);
+    DepResolution.install(map(mdA100, mdB, mdC, mdA110, mdA200, mdD100, mdD200), map(mdB), tml, false);
+    Assert.assertEquals(Json.encodePrettily(tml), 3, tml.size());
+    assertDisable(tml, 0, mdB);
+    assertEnable(tml, 1, mdA200);
+    assertEnable(tml, 2, mdD200);
```

Comment:
```suggestion
    assertThat(tml, contains(enable(mdD200), disable(mdB), enable(mdA200)));
```
Matchers.contains checks size, order and each element.

Requires
```
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
```
and
```
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

public static class TenantModuleDescriptorMatcher extends TypeSafeMatcher<TenantModuleDescriptor> {

  private final Action action;
  private final ModuleDescriptor md;
  private final ModuleDescriptor from;

  public TenantModuleDescriptorMatcher(Action action, ModuleDescriptor md, ModuleDescriptor from) {
      this.action = action;
      this.md = md;
      this.from = from;
  }

  @Override
  public void describeTo(final Description description) {
      description.appendText("matches TenantModuleDescriptor(action = " + action + ", id = " + md.getId()
        + ", from = " + (from == null ? null : from.getId()) + ")");
  }

  @Override
  public boolean matchesSafely(final TenantModuleDescriptor tm) {
    return action == tm.getAction() && md.getId() == tm.getId()
        && (from == null ? tm.getFrom() == null : tm.getFrom() == from.getId());
  }

  public static TenantModuleDescriptorMatcher enable(ModuleDescriptor md) {
    return upgrade(md, null);
  }

  public static TenantModuleDescriptorMatcher upgrade(ModuleDescriptor md, ModuleDescriptor from) {
    return new TenantModuleDescriptorMatcher(Action.enable, md, from);
  }

  public static TenantModuleDescriptorMatcher disable(ModuleDescriptor md) {
    return new TenantModuleDescriptorMatcher(Action.disable, md, null);
  }

  public static TenantModuleDescriptorMatcher upToDate(ModuleDescriptor md) {
    return new TenantModuleDescriptorMatcher(Action.uptodate, md, null);
  }
}
```
and `import static` of TenantModuleDescriptorMatcher.*

Reply:
Added. Thx